descriptors: Be able to specify change and receiving in a single descriptor string

achow101 opened the pull request:
It is convenient to have a descriptor which specifies both receiving and change addresses in a single string. However, as discussed in https://github.com/bitcoin/bitcoin/issues/17190#issuecomment-895515768, it is not feasible to use a generic multipath specification like BIP 88 due to combinatorial blow up and that it would result in unexpected descriptors.

To resolve that problem, this PR proposes a targeted solution which allows only a single pair of 2 derivation indexes to be inserted in the place of a single derivation index. So instead of two descriptor `wpkh(xpub.../0/0/*)` and `wpkh(xpub.../0/1/*)` to represent receive and change addresses, this could be written as `wpkh(xpub.../0/<0;1>/*)`. The multipath specifier is of the form `<NUM;NUM>`. Each `NUM` can have its own hardened specifier, e.g. `<0;1h>` is valid. The multipath specifier can also only appear in one path index in the derivation path.

This results in the parser returning two descriptors. The first descriptor uses the first `NUM` in all pairs present, and the second uses the second `NUM`. In our implementation, if a multipath descriptor is not provided, a pair is still returned, but the second element is just `nullptr`.

The wallet will not output the multipath descriptors (yet). Furthermore, when a multipath descriptor is imported, it is expanded to the two descriptors and each imported on its own, with the second descriptor being implicitly for internal (change) addresses. There is no change to how the wallet stores or outputs descriptors (yet).

Note that the path specifier is different from what was proposed. It uses angle brackets and the semicolon because these are unused characters available in the character set and I wanted to avoid conflicts with characters already in use in descriptors.

Closes #17190

Rspigler commented:
Strong concept ACK!

Glad this is getting a fix.  As discussed in 17190 and OP, there's more improvements that can come after this.

mjdietzx commented:
Concept and Approach ACK. I plan on updating the test and docs introduced in #22067 to take advantage of this in a follow up PR. So I will try to do a more thorough review and testing of this after that is merged and I start on the followup based on this branch. (or if it looks like this will get merged first I will do that earlier)

Sjors commented:
> Note that the path specifier is different from what was proposed. It uses angle brackets and the semicolon because these are unused characters available in the character set and I wanted to avoid conflicts with characters already in use in descriptors.

`{0,1}` seems more readable and would at least be a subset of [BIP 88](https://github.com/bitcoin/bips/blob/master/bip-0088.mediawiki). The conflict is with taproot descriptors, which use `{` and `,` for nesting the scripts.

As an example `tr([00000000/1']xpub/{0,1}/*,{pk([00000000/2']xpub/{0,1}/*),pk([00000000/3']xpub/{0,1}/*)})` doesn't look terrible to me. How much does it really complicate the parsing code? They're always either inside `()` of a descriptor function like `pkh()` or within `[]` of an origin (if we even allow that).

Allowing a mix of hardened and unhardened like `<0;1h>` seems unnecessary to me and might make alternative implementations more complicated (though perhaps it's trivial; I haven't tried).

achow101 commented:
> The conflict is with taproot descriptors, which use `{` and `,` for nesting the scripts.

The conflict is actually with the `,` in a variety of descriptors. The particular issue I had run into was using `,` in a `multi()` descriptor. Since `,` is used to separate the components of `multi()`, the parser would look for the next `,` for the next item to parse. If we use `,` for the multipath specifier, then it would run into that and parse too little. The parser could be changed to not do that, but that was more work than to just use unused characters. `{` did not cause any issues, but to make it easier for other implementations which may be parsing differently, I decided to go with a different character as well.

mjdietzx commented:
In my code I actually have:

```python
result = multisig.importdescriptors([
                {
                    "desc": descriptor["descriptor"],
                    "active": True,
                    "timestamp": "now",
                },
            ])
            assert all(r["success"] for r in result)
```

And that doesn't catch any errors. I will push my branch shortly and reply w/ it

mjdietzx commented:
I had an error in my tests - it seems everything in this PR is working properly.

This works:
```python
result = multisig.importdescriptors([
  {
    "desc": descsum_create(f"wsh(sortedmulti({self.M},{f'/<0;1>/*,'.join(xpubs)}/<0;1>/*))"),
    "active": True,
    "timestamp": "now",
  },
])
```

re the error I posted eariler, I was trying to do this (using `getdescriptorinfo` instead of `descsum_create`):
```python
descriptor = multisig.getdescriptorinfo(f"wsh(sortedmulti({self.M},{f'/<0;1>/*,'.join(xpubs)}/<0;1>/*))")
result = multisig.importdescriptors([
    {
        "desc": descriptor["descriptor"],
        "active": True,
        "timestamp": "now",
    },
])
```

But I was not importing or doing anything with `descriptor["descriptor_change"]`, hence the (obvious in hindsight) error.

---

We used `getdescriptorinfo` instead of `descsum_create` so `test/functional/wallet_multisig_descriptor_psbt.py` will only use node functions without an additional script for the checksum (see review of #22067 for more info). I wonder if this is useful user feedback at all @achow101? I'm specifically wondering about the interface of `getdescriptorinfo` now, and if it would make sense to return something like:
```js
{
  descriptor // still the multipath descriptor, similar to output of descsum_create
  // now the broken out descriptors:
  descriptor_receiving
  descriptor_change
 // ... and the other fields are unchanged
}
```

mjdietzx reviewed (commented):
Getting very close to signing off on this PR, just a few minor things/questions

prusnak commented:
For the record: recently we updated the Trezor stack (Firmware + Suite) to support Taproot and we used descriptors with ranges from this proposal for accounts, e.g. `tr([5c9e228d/86'/0'/0']xpub6Bw885JisRbcKmowfBvMmCxaFHodKn1VpmRmctmJJoM8D4DzyP4qJv8ZdD9V9r3SSGjmK2KJEDnvLH6f1Q4HrobEvnCeKydNvf1eir3RHZk/<0;1>/*)#4swej4wz`

Rspigler commented:
Would like to test but don't know how, as this PR doesn't yet change how descriptors are stored or outputted. Any ideas?

Sjors commented on src/rpc/misc.cpp:
```diff
@@ -170,6 +170,7 @@ static RPCHelpMan getdescriptorinfo()
                 RPCResult::Type::OBJ, "", "",
                 {
                     {RPCResult::Type::STR, "descriptor", "The descriptor in canonical form, without private keys"},
+                    {RPCResult::Type::STR, "descriptor_change", "The change descriptor in canonical form, without private keys. Only if the provided descriptor specifies derivation paths for both receiving and change."},
```
> `descriptor` cannot currently be the multipath descriptor as generating the string output for one is not yet implemented. I'm not sure on a good approach for doing that yet.

Mmm, that has me a bit worried. If a user imports a multipath descriptor, I think we should remember it in that form. And if we embrace multipath descriptors, we should probably also use them for new wallets.

prusnak commented:
Should `<NUM;NUM;NUM;NUM>` also work or we only aim to support only tuples of exactly two values?

craigraw commented:
ACK.

Successfully tested `getdescriptorinfo`, `deriveaddresses` and `importdescriptors` against singlesig and multisig multipath descriptors generated by Sparrow. Wallet balances, addresses and checksums all match. Sparrow uses multipath descriptors since Nov 2021.

Rspigler commented:
tACK commit 057511d35375bada45a9b71b7e06db1edcb34091 

Computing checksum of the multipath descriptor:

`getdescriptorinfo "wpkh([2fb36ae6/84'/0'/0']xpub6Bpt6XP21UMYtMC7xjKsMYorev7si6TY3QfqFf3jLP6LvTcdV66DGnGcW6TZWduhMBKBqC8XykFj5zKPXvS4TasoeHYsozqqdfH1ZZ7fZMK/<0;1>/*)"`

> {
>   "descriptor": "wpkh([2fb36ae6/84'/0'/0']xpub6Bpt6XP21UMYtMC7xjKsMYorev7si6TY3QfqFf3jLP6LvTcdV66DGnGcW6TZWduhMBKBqC8XykFj5zKPXvS4TasoeHYsozqqdfH1ZZ7fZMK/0/*)#ljnxrhyu",
>   "descriptor_change": "wpkh([2fb36ae6/84'/0'/0']xpub6Bpt6XP21UMYtMC7xjKsMYorev7si6TY3QfqFf3jLP6LvTcdV66DGnGcW6TZWduhMBKBqC8XykFj5zKPXvS4TasoeHYsozqqdfH1ZZ7fZMK/1/*)#wxk87z5y",
>   "checksum": "ava8ef96",
>   "isrange": true,
>   "issolvable": true,
>   "hasprivatekeys": false
> }

`importdescriptors "[{\"desc\":\"wpkh([2fb36ae6/84'/0'/0']xpub6Bpt6XP21UMYtMC7xjKsMYorev7si6TY3QfqFf3jLP6LvTcdV66DGnGcW6TZWduhMBKBqC8XykFj5zKPXvS4TasoeHYsozqqdfH1ZZ7fZMK/<0;1>/*)#ava8ef96\",\"range\":[0,1000],\"timestamp\":1657145357}]"`

> [
>   {
>     "success": true
>   }
> ]


Unrelated to this specific PR, was `importdescriptor`, since you have to escape all the double quotes, and there's no mention of this in the console documentation (and there's also a discrepancy in the `timestamp` spacing).  Putting this here for notes, I might do this myself

MarcoFalke commented on src/wallet/rpc/spend.cpp:
```diff
@@ -639,12 +639,14 @@ void FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& fee_out,
-                desc->Expand(0, desc_out, scripts_temp, desc_out);
-                coinControl.m_external_provider = Merge(coinControl.m_external_provider, desc_out);
+                for (auto& desc : descs) {
+                    desc->Expand(0, desc_out, scripts_temp, desc_out);
```
Wouldn't `desc_out` need to be moved inside the scope of this for loop to avoid accumulating duplicate descriptors?

achow101 replied:
It doesn't matter since it all gets accumulated into into `coinControl.m_external_provider` anyways.

fanquake commented:
In [the fuzzing job](https://cirrus-ci.com/task/4965086438621184): 
```bash
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x67,0x65,0x74,0x64,0x65,0x73,0x63,0x72,0x69,0x70,0x74,0x6f,0x72,0x69,0x6e,0x66,0x6f,0x5c,0xa,0x77,0x73,0x68,0x28,0x75,0x3a,0x30,0x29,0x2c,0x50,0x75,0x75,0x6c,0x3a,0x31,0x29,
getdescriptorinfo\\\012wsh(u:0),Puul:1)
artifact_prefix='./'; Test unit written to ./oom-15b296231df90f4407beea7c9de1b8c129658fae
Base64: Z2V0ZGVzY3JpcHRvcmluZm9cCndzaCh1OjApLFB1dWw6MSk=
SUMMARY: libFuzzer: out-of-memory
```

achow101 commented:
It's unclear to me what the fuzzer is OOM'ing on.

achow101 commented:
I think I fixed the fuzzer issue.

darosior reviewed (commented):
Concept ACK.

darosior reviewed (commented):
ACK on most of the core logic. Just found a small mistake in the parsing of the multipath step in the derivation path. I also have a couple questions on the modifications to the RPC interface. The rest is nits, feel free to ignore them. Nice set of unit tests!

One almost-nit but important for posterity IMO is that all your commit messages mention the previous iteration of the BIP that only allowed for 2 paths (change and receive) and should be adapted to mention "multiple" paths to avoid confusing a reader.

As i've already mentioned in the BIP review last year (https://github.com/bitcoin/bips/pull/1354#discussion_r985316649), restricting it to two paths would remove a lot of complexity without preventing any known usecase. Happy to keep it as it is though, i believe the code is correct.

I've compared this to my Rust (for rust-miniscript, see https://github.com/rust-bitcoin/rust-miniscript/pull/470) and Python (https://github.com/darosior/python-bip380/pull/21) implementations, suggesting a couple more unit test cases but without finding any issue.

I also wrote a more efficient descriptor parsing fuzzer that i grew to more branch coverage than the existing `descriptor_parse` target. I also used a dictionary with the multipath derivation step syntax on both the existing and my new targets. All the multipath parsing code was covered and i didn't find any crash.

achow101 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1102,25 +1107,34 @@ static UniValue ProcessImportDescriptor(ImportData& import_data, std::map<CKeyID
+    for (size_t j = 0; j < parsed_descs.size(); ++j) {
+        const auto& parsed_desc = parsed_descs.at(j);
+        bool desc_internal = internal.has_value() && internal.value();
+        if (parsed_descs.size() == 2) {
+            desc_internal = j == 1;
```
Since the main point of this entire proposal is to make it easier to import receiving and change as a single item, I think it makes sense to special case these.

darosior replied:
Then let's at least document in `importdescriptors` and `importmulti` that for any multipath descriptor with 2 paths, the second will be imported as internal?

achow101 replied:
Added a help text.

achow101 commented on src/wallet/test/psbt_wallet_tests.cpp:
```diff
@@ -20,8 +20,10 @@ static void import_descriptor(CWallet& wallet, const std::string& descriptor)
-    std::unique_ptr<Descriptor> desc = Parse(descriptor, provider, error, /* require_checksum=*/ false);
-    assert(desc);
+    auto descs = Parse(descriptor, provider, error, /* require_checksum=*/ false);
+    assert(!descs.empty());
+    assert(descs.size() == 1);
```
Removed

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -268,15 +309,13 @@ static RPCHelpMan deriveaddresses()
                 std::tie(range_begin, range_end) = ParseDescriptorRange(request.params[1]);
             }
 
+
```
Removed

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1780,8 +1780,9 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
-        std::unique_ptr<Descriptor> desc = Parse(desc_str, keys, error, false);
-        WalletDescriptor w_desc(std::move(desc), creation_time, 0, 0, 0);
+        std::vector<std::unique_ptr<Descriptor>> descs = Parse(desc_str, keys, error, false);
+        assert(descs.size() == 1); // It shouldn't be possible to have an invalid or multipath descriptor
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1647,20 +1758,52 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                        vec.emplace_back(vec.at(0)->Clone());
+                    }
+                } else if (vec.size() != *num_multipath) {
+                    error = strprintf("Miniscript: Multipath subscripts have mismatched lengths");
```
Done

achow101 commented on src/test/descriptor_tests.cpp:
```diff
@@ -424,6 +464,259 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
+    CheckUnparsable("tr(xprv9s21ZrQH143K2Zu2kTVKcQi9nKhfgJUkYqG73wXsHuhATm1wkt6kcSZeTYEw2PL7krZtJopEYDvBdYWdAai3n3TWUTCVfHvPHqTYJv7smYe/<6;7;8;9>/*,{pk(xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc/<1;2;3>/0/*),pk(xprv9s21ZrQH143K3jUwNHoqQNrtzJnJmx4Yup8NkNLdVQCymYbPbJXnPhwkfTfxZfptcs3rLAPUXS39oDLgrNKQGwbGsEmJJ8BU3RzQuvShEG4/0/0/<3; ...
+    CheckUnparsable("tr(xprv9s21ZrQH143K2Zu2kTVKcQi9nKhfgJUkYqG73wXsHuhATm1wkt6kcSZeTYEw2PL7krZtJopEYDvBdYWdAai3n3TWUTCVfHvPHqTYJv7smYe/<6;7>/*,{pk(xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc/<1;2;3>/0/*),pk(xprv9s21ZrQH143K3jUwNHoqQNrtzJnJmx4Yup8NkNLdVQCymYbPbJXnPhwkfTfxZfptcs3rLAPUXS39oDLgrNKQGwbGsEmJJ8BU3RzQuvShEG4/0/0/<3;4;5> ...
+    CheckUnparsable("sh(multi(2,xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc/<1;2;3>/0/*,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/0/*,xprv9s21ZrQH143K3jUwNHoqQNrtzJnJmx4Yup8NkNLdVQCymYbPbJXnPhwkfTfxZfptcs3rLAPUXS39oDLgrNKQGwbGsEmJJ8BU3RzQuvShEG4/0/0/<3;4>/*))", ...
+
```
Added tests as suggested

achow101 commented on src/test/descriptor_tests.cpp:
```diff
@@ -424,6 +464,259 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
+                {{6, 0}, {6, 1}, {6, 2}, {2, 0, 0}, {2, 0, 1}, {2, 0, 2}, {0, 0, 4, 0}, {0, 0, 4, 1}, {0, 0, 4, 2}},
+                {{6, 0}, {6, 1}, {6, 2}, {3, 0, 0}, {3, 0, 1}, {3, 0, 2}, {0, 0, 5, 0}, {0, 0, 5, 1}, {0, 0, 5, 2}},
+            }
+    );
```
Added as suggested

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -256,11 +270,14 @@ static RPCHelpMan deriveaddresses()
             }
-
+            if (descs.size() > 1) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Descriptor specifying both receiving and change derivation paths are not allowed");
```
Changed

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1647,20 +1758,52 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                // No keys, return single descriptor
+                ret.emplace_back(std::make_unique<MiniscriptDescriptor>(std::vector<std::unique_ptr<PubkeyProvider>>(), std::move(node)));
+                return ret;
+            }
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1153,29 +1153,64 @@ enum class ParseScriptContext {
  * @returns false if parsing failed
  **/
-[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out, bool& apostrophe, std::string& error)
+[[nodiscard]] bool ParseKeyPath(const std::vector<Span<const char>>& split, std::vector<KeyPath>& out, bool& apostrophe, std::string& error, bool allow_multipath = false)
```
Done

achow101 commented on src/script/descriptor.h:
```diff
@@ -156,7 +156,7 @@ struct Descriptor {
  * else is wrong, `nullptr` is returned.
  */
-std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out, std::string& error, bool require_checksum = false);
+std::vector<std::unique_ptr<Descriptor>> Parse(const std::string& descriptor, FlatSigningProvider& out, std::string& error, bool require_checksum = false);
```
Done

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -197,16 +201,25 @@ static RPCHelpMan getdescriptorinfo()
 
             UniValue result(UniValue::VOBJ);
-            result.pushKV("descriptor", desc->ToString());
+            result.pushKV("descriptor", descs.at(0)->ToString());
```
Added a sentence in the doc.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1153,29 +1153,64 @@ enum class ParseScriptContext {
+                error = "Multiple multipath key path specifiers found";
+                return false;
+            }
+            to_process = Split(item.subspan(1, item.size() - 2), ';');
```
Added a check and the test.

darosior commented:
This needs rebase for CI to pass.

darosior commented:
The descriptor on which it crashed is `wsh(0)`. This is a bug in the Miniscript logic, exposed by the new check but not introduced in this PR. A `0` script should not be considered sane.

darosior commented:
I think the `CHECK_NONFATAL` that we don't parse Miniscript descriptors with an empty set of keys can be removed here, No need to hold off this PR and it was added in https://github.com/bitcoin/bitcoin/pull/27997 which fixes the bug it uncovered.

darosior commented:
re-ACK 33a2f6dc3c2512129df7882586acb3b3ddce62cb

darosior commented:
re-ACK 8f6eeaab38

darosior commented:
Happy 2 years birthday, #22838! :birthday:

darosior reviewed (changes requested):
The fuzz target is crashing because you updated `is_solvable` in `TestDescriptor` to be an `optional` but didn't update `!is_solvable` to `!*is_solvable`.

```diff
diff --git a/src/test/fuzz/descriptor_parse.cpp b/src/test/fuzz/descriptor_parse.cpp
index 2dd55fda84..4a2ad65fd0 100644
--- a/src/test/fuzz/descriptor_parse.cpp
+++ b/src/test/fuzz/descriptor_parse.cpp
@@ -151,7 +151,7 @@ static void TestDescriptor(const Descriptor& desc, FlatSigningProvider& sig_prov
     const auto max_sat_nonmaxsig{desc.MaxSatisfactionWeight(true)};
     const auto max_elems{desc.MaxSatisfactionElems()};
     // We must be able to estimate the max satisfaction size for any solvable descriptor (but combo).
-    const bool is_nontop_or_nonsolvable{!is_solvable || !desc.GetOutputType()};
+    const bool is_nontop_or_nonsolvable{!*is_solvable || !desc.GetOutputType()};
     const bool is_input_size_info_set{max_sat_maxsig && max_sat_nonmaxsig && max_elems};
     assert(is_input_size_info_set || is_nontop_or_nonsolvable);
 }
```

darosior commented:
> Successfully tested `getdescriptorinfo`, `deriveaddresses` and `importdescriptors` against singlesig and multisig multipath descriptors generated by Sparrow. Wallet balances, addresses and checksums all match. Sparrow uses multipath descriptors since Nov 2021.

@craigraw would you mind testing against the latest state of this PR to help move things forward here?

craigraw commented:
> Successfully tested `getdescriptorinfo`, `deriveaddresses` and `importdescriptors` against singlesig and multisig multipath descriptors generated by Sparrow. Wallet balances, addresses and checksums all match. 

Tested the above again against the current state of the PR, no issues found.

achow101 commented:
Fixed the fuzz crash

darosior reviewed (commented):
re-ACK ba32818d09310106094127b64cc2c249c7f723c0

furszy commented:
Added to my review queue. Will come here after finishing with #26728.

darosior commented:
If i re-review this would anyone else review as well? ACKed it multiple times in the past but it didn't get merged because we'd need another person.

It's been a while, i need to review it from scratch so i'm trying to probe if that's worth spending the time.

achow101 commented:
Rebasing for CI

furszy commented on src/script/descriptor.cpp:
```diff
@@ -786,6 +787,11 @@ class DescriptorImpl : public Descriptor
+=======
+
+    virtual std::unique_ptr<DescriptorImpl> Clone() const = 0;
+>>>>>>> 657176b834d (descriptors: Add DescriptorImpl::Clone)
```
Rebase issues in ed00e60c8e8 that are removed on 06caba093a.

achow101 replied:
Fixed

mjdietzx commented:
Tested ACK cb41e26eef3a92a7704a0e01eb9b272f7ed78eeb. Note: I did not do a thorough code review

What I've tested:
- `getdescriptorinfo`, `importdescriptors`, `listdescriptors` with a miniscript wallet similar to [#29156](https://github.com/bitcoin/bitcoin/pull/29156). All works as expected
- import this wallet from file to Coldcard Mk4 (Edge FW 6.2.2 w/ miniscript support)
  - `<0;1>` notation works properly as Coldcard has already adopted this convention
  - Coldcard also signs psbt properly
- I have two PRs which would benefit from this improvement. I rebased [#29156](https://github.com/bitcoin/bitcoin/pull/29156) on top of this. This diff simplifies and passes:
```diff
diff --git a/test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py b/test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py
index 4f50ae191c..e5f58eb4d2 100755
--- a/test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py
+++ b/test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py
@@ -32,35 +32,26 @@ class WalletMiniscriptDecayingMultisigDescriptorPSBTTest(BitcoinTestFramework):
         self.skip_if_no_sqlite()
 
     @staticmethod
-    def _get_xpub(wallet, internal):
+    def _get_xpub(wallet):
         """Extract the wallet's xpubs using `listdescriptors` and pick the one from the `pkh` descriptor since it's least likely to be accidentally reused (legacy addresses)."""
-        pkh_descriptor = next(filter(lambda d: d["desc"].startswith("pkh(") and d["internal"] == internal, wallet.listdescriptors()["descriptors"]))
+        pkh_descriptor = next(filter(lambda d: d["desc"].startswith("pkh("), wallet.listdescriptors()["descriptors"]))
         # keep all key origin information (master key fingerprint and all derivation steps) for proper support of hardware devices
         # see section 'Key origin identification' in 'doc/descriptors.md' for more details...
         return pkh_descriptor["desc"].split("pkh(")[1].split(")")[0]
 
-    def create_multisig(self, external_xpubs, internal_xpubs):
+    def create_multisig(self, xpubs):
         """The multisig is created by importing the following descriptors. The resulting wallet is watch-only and every signer can do this."""
         self.node.createwallet(wallet_name=f"{self.name}", blank=True, descriptors=True, disable_private_keys=True)
         multisig = self.node.get_wallet_rpc(f"{self.name}")
         # spending policy: `thresh(4,pk(key_1),pk(key_2),pk(key_3),pk(key_4),after(t1),after(t2),after(t3))`
         # IMPORTANT: when backing up your descriptor, the order of key_1...key_4 must be correct!
-        external = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(external_xpubs)}),sln:after({f'),sln:after('.join(map(str, self.locktimes))})))")
-        internal = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(internal_xpubs)}),sln:after({f'),sln:after('.join(map(str, self.locktimes))})))")
-        result = multisig.importdescriptors([
-            {  # receiving addresses (internal: False)
-                "desc": external["descriptor"],
-                "active": True,
-                "internal": False,
-                "timestamp": "now",
-            },
-            {  # change addresses (internal: True)
-                "desc": internal["descriptor"],
-                "active": True,
-                "internal": True,
-                "timestamp": "now",
-            },
-        ])
+        descriptors = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(xpubs)}),sln:after({f'),sln:after('.join(map(str, self.locktimes))})))")
+        result = multisig.importdescriptors([{
+            "desc": desc,
+            "active": True,
+            "internal": "/1/*" in desc,
+            "timestamp": "now",
+        } for desc in descriptors["multipath_expansion"]])
         assert all(r["success"] for r in result)
         return multisig
 
@@ -77,10 +68,10 @@ class WalletMiniscriptDecayingMultisigDescriptorPSBTTest(BitcoinTestFramework):
 
         self.log.info("Create the signer wallets and get their xpubs...")
         signers = [self.node.get_wallet_rpc(self.node.createwallet(wallet_name=f"signer_{i}", descriptors=True)["name"]) for i in range(self.N)]
-        external_xpubs, internal_xpubs = [[self._get_xpub(signer, internal) for signer in signers] for internal in [False, True]]
+        xpubs = [self._get_xpub(signer).replace("/0/*", "/<0;1>/*") for signer in signers]
 
         self.log.info("Create the watch-only decaying multisig using signers' xpubs...")
-        multisig = self.create_multisig(external_xpubs, internal_xpubs)
+        multisig = self.create_multisig(xpubs)
 
         self.log.info("Get a mature utxo to send to the multisig...")
         coordinator_wallet = self.node.get_wallet_rpc(self.node.createwallet(wallet_name=f"coordinator", descriptors=True)["name"])
```

darosior commented:
Will re-review from scratch next week to finally get this over the finish line.

pythcoiner commented:
I'm not sure if it's the intended behavior, but passing the same `NUM` twice (here `<1;1>`) to `importdescriptors` does not fail and ends up with a single change descriptor:

```python
self.log.info("Multipath descriptors: duplicate NUM")
self.nodes[1].createwallet(wallet_name="multipath3", descriptors=True, blank=True)
w_multipath = self.nodes[1].get_wallet_rpc("multipath3")
self.test_importdesc({"desc": descsum_create(f"wpkh({xpriv}/<1;1>/*)"),
                      "active": True,
                      "range": 10,
                      "timestamp": "now"},
                      success=True,
                      wallet=w_multipath)
result = w_multipath.listdescriptors()
self.log.info(f"{result=}")
assert_equal(len(result["descriptors"]), 1)

assert_equal(result["descriptors"][0]["desc"],"wpkh(tpubD6NzVbkrYhZ4YNXVQbNhMK1WqguFsUXceaVJKbmno2aZ3B6QfbMeraaYvnBSGpV3vxLyTTK9DYT1yoEck4XUScMzXoQ2U2oSmE2JyMedq3H/1/*)#hf9h03qd" )
assert_equal(result["descriptors"][0]["internal"], True)


```

achow101 commented:
> I'm not sure if it's the intended behavior, but passing the same `NUM` twice to `importdescriptors` does not fail and ends up with a single _change_ descriptor:

Good catch.

The BIP does not specify either way yet, but I don't think there's any reason to allow duplicates here, so I've implemented that and added a test.

pythcoiner commented:
note: overlapping multipath are allowed
```python
        self.test_importdesc({"desc": descsum_create(f"wsh(and_v(v:pk({xpriv}/<0;1>/*),or_d(pk({xpriv}/<1;2>/*),older(1000))))"),
                              "active": True,
                              "range": 10,
                              "timestamp": "now"},
                              success=True,
                              wallet=w_multipath)
```
but I don't see how it can be harmfull as it ends up in valid miniscript descriptor

pythcoiner commented:
importing a descriptor w/ `"internal": false` fail because checks [here](https://github.com/bitcoin/bitcoin/blob/00337ef0e115f8bd8c1ede425f21ae7a1b6d30de/src/wallet/rpc/backup.cpp#L1073) and [here](https://github.com/bitcoin/bitcoin/blob/00337ef0e115f8bd8c1ede425f21ae7a1b6d30de/src/wallet/rpc/backup.cpp#L1476) are against if the key exist, not against its value.

```shell
~/cpp/bitcoin/src pr22838 *1 ?3 ❯ bitcoin-cli -regtest -rpcwallet=liana_recovery importdescriptors "[{\"desc\": \"wsh(or_d(pk([8d757e4c/48'/1'/0'/2']tpubDEmyHnnFiU6hyYrKsnJjfYmeqabZjG69mh1fosrYjf8XFaeppCkLKuc6JefqyZKHxNtgZwJ5KuJ51yLLSt66zc1iyFnL9J8gpN7kSfLvnvG/<0;1>/*),and_v(v:pkh([8d757e4c/48'/1'/0'/2']tpubDEmyHnnFiU6hyYrKsnJjfYmeqabZjG69mh1fosrYjf8XFaeppCkLKuc6JefqyZKHxNtgZwJ5KuJ51yLLSt66zc1iyFnL9J8gpN7kSfLvnvG/<2;3>/*),older(65535))))#87grqs3t\", \"range\": [0,10000], \"timestamp\": \"now\", \"active\": true, \"internal\": false}]"
[
  {
    "success": false,
    "error": {
      "code": -5,
      "message": "Cannot have multipath descriptor while also specifying 'internal'"
    }
  }
]
```

darosior reviewed (approved):
ACK 00337ef0e115f8bd8c1ede425f21ae7a1b6d30de

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1414,34 +1414,76 @@ enum class ParseScriptContext {
+                return false;
+            }
+            p |= ((uint32_t)hardened) << 31;
+            if (std::find(multipath_elems.begin(), multipath_elems.end(), p) != multipath_elems.end()) {
```
We need the vector since order has to be preserved, and since the number of elements should be small, I don't think the performance impact is significant. Also, `std::find` is `O(n)`.

darosior replied:
It's `O(n)` but it's called in a loop over `n`.. So it effectively does `(n^2 - n)/2` operations, which is `O(n^2)`.

pythcoiner commented:
ACK 00337ef
I've written some [tests](https://github.com/bitcoin/bitcoin/commit/f6cdf9bec82533e849f6156f5593dc9ceb688306) and tested import descriptors produced by liana wallet

darosior commented:
ACK a512245e217199208807214b09bdb256362d08af

pythcoiner commented:
ACK a512245

darosior commented:
I'd really like to get this in before the feature freeze for v28. I think it's in good shape and it's regaining momentum.

@mjdietzx would you be able to go over the diff since your last review soon? It didn't change much since your ACK.
@furszy are you still interested in reviewing this work?

mjdietzx reviewed (approved):
ACK a512245e217199208807214b09bdb256362d08af

Great job and thank you for implementing this!

furszy commented:
> @furszy are you still interested in reviewing this work?

Yes. Starting with it. Thanks for the ping.

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1414,34 +1414,76 @@ enum class ParseScriptContext {
+                    apostrophe = last == '\'';
+                }
+            }
+            uint32_t p;
+            if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p)) {
+                error = strprintf("Key path value '%s' is not a valid uint32", std::string(elem.begin(), elem.end()));
+                return false;
+            } else if (p > 0x7FFFFFFFUL) {
+                error = strprintf("Key path value %u is out of range", p);
+                return false;
+            }
+            p |= ((uint32_t)hardened) << 31;
+            if (std::find(multipath_elems.begin(), multipath_elems.end(), p) != multipath_elems.end()) {
+                error = strprintf("Duplicated key path value %u in multipath specifier", p);
+                return false;
             }
+            multipath_elems.push_back(p);
         }
-        uint32_t p;
-        if (!ParseUInt32(std::string(elem.begin(), elem.end()), &p)) {
-            error = strprintf("Key path value '%s' is not a valid uint32", std::string(elem.begin(), elem.end()));
-            return false;
-        } else if (p > 0x7FFFFFFFUL) {
-            error = strprintf("Key path value %u is out of range", p);
-            return false;
+        if (multipath_elems.size() == 1) {
+            for (auto& kp : out) {
+                kp.push_back(multipath_elems.at(0));
+            }
+            continue;
+        }
+        while (out.size() != multipath_elems.size()) {
+            out.push_back(out.at(0));
+        }
+        for (size_t j = 0; j < multipath_elems.size(); ++j) {
+            out[j].push_back(multipath_elems.at(j));
         }
-        out.push_back(p | (((uint32_t)hardened) << 31));
     }
+
```
I didn't find this implementation so friendly and ended up reimplementing the function in a more readable way (at least for me) https://github.com/furszy/bitcoin-core/commit/84bbef2de7f4d3a697b28f6b9177e20dc974b8cc. All yours if you like it. It contain other small code improvements too.

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1556,7 +1600,9 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkey(uint32_t key_exp_index,
     assert(fpr_bytes.size() == 4);
     std::copy(fpr_bytes.begin(), fpr_bytes.end(), info.fingerprint);
-    if (!ParseKeyPath(slash_split, info.path, apostrophe, error)) return {};
+    std::vector<KeyPath> path;
+    if (!ParseKeyPath(slash_split, path, apostrophe, error)) return {};
+    info.path = path.at(0);
```
In fa53d191a02:

Could be explicit here. The general `slash_split` variable naming doesn't really help. 
```suggestion
    // Extract origin path elements
    std::vector<KeyPath> path;
    if (!ParseKeyPath(slash_split, path, apostrophe, error, /**allow_multipath=**/false)) return {};
    info.path = path.at(0); // multipath specifier disabled for origin information
```
.
Can also remove the `allow_multipath` default value from the `ParseKeyPath` signature.

achow101 replied:
Removed the default value. The variable names weren't added/changed in this PR.

furszy reviewed (commented):
Thinking out loud here.
Since we are not expanding the multipath specifier to cover all possibilities due to a potentially large combinatorial space and are instead returning only those matching the Ith term, I believe we could enable value duplication within the new specifier for multisig descriptors. For example `multi(1, pkA/<0;1;2>/0, pkB/<1;2;1>/0)` could allow users to retrieve `multi(1, pkA/0/0, pkB/1/0)`, `multi(1, pkA/1/0, pkB/2/0)`, and `multi(1, pkA/2/0, pkB/1/0)`. This would play well with the multipath values matching lengths requirement.

achow101 commented on test/functional/rpc_getdescriptorinfo.py:
```diff
@@ -19,10 +19,14 @@ def set_test_params(self):
+            assert_equal(info["descriptor"], descsum_create(expanded_descs[0]))
+            assert_equal(info["multipath_expansion"], [descsum_create(x) for x in expanded_descs])
+        else:
+            assert "multipath_expansion" not in info
```
Good catch, added it back in.

darosior reviewed (commented):
re-ACK 8208454eae7484d34a162ffa7701157e56e3cb80

Although i agree it's more readable now it's unfortunate to ask to rewrite part of the parsing logic only for style purpose after it had 3 ACKs and underwent significant fuzzing.

That said i didn't find any issue and neither did my fuzzer (for now). I'll keep fuzzing this PR rebased on master for a day.

furszy reviewed (commented):
Dropping another out loud thinking:
The multipath specifier can be used in a way it degrades wallet performance for no reason if it is used in the last segment of single key constructions. For instance, a `pkh(xpub../0/<0;1;2;3;4;5;6;7;8;9;10,...>)` imports 10 different descriptors into the wallet while the same can be expressed using a single ranged descriptor `pkh(xpub../0/*)` with range 0-10.

Topic here is whether we should disable it for top-level single key functions, alert users about the degraded performance or do nothing. I am not suggesting disabling it for all functions because the functionality is useful for multi-key constructions. Thoughts?

furszy reviewed (commented):
still reviewing, only nits so far.

> Although i agree it's more readable now it's unfortunate to ask to rewrite part of the parsing logic only for style purpose after it had 3 ACKs and underwent significant fuzzing.

I think it depends on the proposed changes. If readability is improved substantially (bolding substantially), it helps maintenance and might also uncover logical issues that the fuzzer might not easily encounter. I don't think we should settle anything in stone because of a good fuzzing coverage. We should have enough time for fuzzing this properly during the freeze period.

Yet, thats my general opinion. I understand that it is annoying to re-review code you've checked dozens of times over the past years and want to move on once for all. So.. will just continue reviewing 🚜.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1410,37 +1410,94 @@ enum class ParseScriptContext {
+            for (const auto& num : nums) {
+                const auto& op_num = ParseKeyPathNum(num, apostrophe, error);
+                if (!op_num) return false;
+                if (std::find(multipath_values.begin(), multipath_values.end(), *op_num) != multipath_values.end()) {
```
Done, since I'm making changes anyways.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1410,37 +1410,94 @@ enum class ParseScriptContext {
+        for (size_t i = 0; i < multipath_values.size(); i++) {
+            KeyPath branch_path = path;
+            branch_path[*multipath_segment_index] = multipath_values[i];
+            out.emplace_back(branch_path);
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1728,118 +1751,159 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+        size_t num_multipath = std::max_element(providers.begin(), providers.end(),
+                [](const std::vector<std::unique_ptr<PubkeyProvider>>& a, const std::vector<std::unique_ptr<PubkeyProvider>>& b) {
+                    return a.size() < b.size();
+                })->size();
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -1728,118 +1751,159 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                })->size();
+
+        // Make sure all vecs are of the same length, or exactly length 1
+        // For length 1 vectors, clone subdescs until vector is the same length
```
Done

achow101 committed:
descriptors: Add PubkeyProvider::Clone

achow101 committed:
descriptors: Add DescriptorImpl::Clone

achow101 committed:
descriptors: Change ParseScript to return vector of descriptors

To prepare for returning multipath descriptors which will be a shorthand
for specifying multiple descriptors, change ParseScript's signature to return a
vector.

achow101 committed:
descriptors: Have ParseKeypath handle multipath specifiers

Multipath specifiers are derivation path indexes of the form `<i;j;k;...>`
used for specifying multiple derivation paths for a descriptor.
Only one multipath specifier is allowed per PubkeyProvider.
This is syntactic sugar which is parsed into multiple distinct descriptors.
One descriptor will have all of the `i` paths, the second all of the `j` paths,
the third all of the `k` paths, and so on.

ParseKeypath will always return a vector of keypaths with the same size
as the multipath specifier. The callers of this function are updated to deal
with this case and return multiple PubkeyProviders. Their callers have
also been updated to handle vectors of PubkeyProviders.

Co-Authored-By: furszy <matiasfurszyfer@protonmail.com>

achow101 committed:
descriptors: Change Parse to return vector of descriptors

When given a descriptor which contins a multipath derivation specifier,
a vector of descriptors will be returned.

achow101 committed:
tests: Add unit tests for multipath descriptors

achow101 committed:
tests: Multipath descriptors for getdescriptorinfo

achow101 committed:
rpc: Have deriveaddresses derive receiving and change

When given a multipath descriptor, derive all of the descriptors.
The derived addresses will be returned in an object
consisting of multiple arrays. For compatibility, when given a single path
descriptor, the addresses are provided in a single array as before.

achow101 committed:
tests: Multipath descriptors for scantxoutset and deriveaddresses

achow101 committed:
wallet: Move internal to be per key when importing

Instead of applying internal-ness to all keys being imported at the same
time, apply it on a per key basis. So each key that is imported will
carry with it whether it is for the change keypool.

achow101 committed:
rpc: Allow importmulti to import multipath descriptors correctly

Multipath descriptors will be imported as multiple separate descriptors.
When there are exactly 2 multipath items, the first descriptor will be
for receiving addreses, and the second for change
addresses. When importing a multipath descriptor, 'internal' cannot be
specified.

achow101 committed:
wallet, rpc: Allow importdescriptors to import multipath descriptors

Multipath descriptors will be imported as multiple separate descriptors.
When there are 2 multipath items, the first descriptor will be for receiving
addresses and the second for change. This mirrors importmulti.

achow101 committed:
tests: Test importing of multipath descriptors

Test that both importmulti and importdescriptors behave as expected when
importing a multipath descriptor.

achow101 committed:
doc: Mention multipath specifier

furszy commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1596,6 +1615,7 @@ RPCHelpMan importdescriptors()
 {
     return RPCHelpMan{"importdescriptors",
                 "\nImport descriptors. This will trigger a rescan of the blockchain based on the earliest timestamp of all descriptors being imported. Requires a new wallet backup.\n"
+            "When importing descriptors with multipath key expressions, if the multipath specifier contains exactly two elements, the descriptor produced from the second elements will be imported as an internal descriptor.\n"
```
In 7ca8b5b5c:

This could be the default behavior but when the "internal" argument is provided, I think all descriptors should follow that argument (all internal or all external if the flag is set). Then, on a follow-up, we could make "internal" accept an array, allowing users to precisely determine which descriptor in the multipath is internal and which is not.

achow101 replied:
The intended and expected usage is for 2 multipath elements for receiving and change. I think diverging from that for default behavior will be confusing, especially since multipath descriptors are in use by other software and use this pattern. I'm fine with a followup that allows the user to override that and specify which should be internal.

furszy reviewed (commented):
Almost finish, left a question https://github.com/bitcoin/bitcoin/pull/22838#discussion_r1709846109. The rest are nits.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -2108,7 +2241,12 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
         KeyParser parser(/* out = */nullptr, /* in = */&provider, /* ctx = */script_ctx);
         auto node = miniscript::FromScript(script, parser);
         if (node && node->IsSane()) {
-            return std::make_unique<MiniscriptDescriptor>(std::move(parser.m_keys), std::move(node));
+            std::vector<std::unique_ptr<PubkeyProvider>> keys;
+            keys.reserve(parser.m_keys.size());
+            for (auto& key : parser.m_keys) {
+                keys.emplace_back(std::move(key.at(0)));
+            }
+            return std::make_unique<MiniscriptDescriptor>(std::move(keys), std::move(node));
```
I've updated the comment for `m_keys` to hopefully make this clearer.

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -250,11 +264,14 @@ static RPCHelpMan deriveaddresses()
             }
-
+            if (descs.size() > 1) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Descriptor with mulitpath derivation path specifiers are not allowed");
```
Done

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -280,36 +323,18 @@ static RPCHelpMan deriveaddresses()
                 }
+                return ret;
+            } else {
+                return addresses;
```
Done

furszy reviewed (commented):
Code review ACK a0abcbd

darosior commented:
re-ACK a0abcbd3822bd17a1d73c42ccd5b040a150b0501

Running a fuzzer for a couple days on the previous push (rebased on master to get the speedups from #30197) and with a dictionary specifically targeting the multipath expressions did not uncover anything.

glozow reviewed (commented):
light code review ACK a0abcbd3822

I'm unfamiliar with the descriptor code, but the refactoring commits look like pure refactoring, behavior change commits look correct, RPC changes seem intuitive, and tests seem to cover things when I introduce mutations.

achow101 commented on src/rpc/output_script.cpp:
```diff
@@ -234,17 +268,29 @@ static RPCHelpMan deriveaddresses()
-                {RPCResult::Type::STR, "address", "the derived addresses"},
-            }
+        {
+            RPCResult{"for single derivation descriptors",
+                RPCResult::Type::ARR, "", "",
+                {
+                    {RPCResult::Type::STR, "address", "the derived addresses"},
+                }
+            },
+            RPCResult{"for multipath descriptors",
```
I expect single path to still be most used and I thought that returning an array for that would be confusing.

glozow replied:
Makes sense

achow101 commented:
@pythcoiner @mjdietzx Can you please re-review?

mjdietzx commented:
reACK a0abcbd3822bd17a1d73c42ccd5b040a150b0501

pythcoiner commented:
reACK a0abcbd

jonatack reviewed (commented):
With this merge, would it be a good idea to update relevant BIPs (possibly 88, 388, 389)?

e.g. https://github.com/bitcoin/bips/blob/master/bip-0088.mediawiki#compatibility

"There is a discussion on path templating for bitcoin script descriptors at https://github.com/bitcoin/bitcoin/issues/17190, which proposes the format <code>xpub...{0,1}/*</code>, of which the <code>{0,1}/*</code> part would correspond to the partial path template in the format of this BIP."
